vault/core.go

```diff
@@ -2939,3 +2947,49 @@ type LicenseState struct {
 	ExpiryTime time.Time
 	Terminated bool
 }
+
+type InFlightRequests struct {
+	InFlightReqMap   *sync.Map
+	InFlightReqCount *uberAtomic.Uint64
+}
+
+type InFlightReqData struct {
+	StartTime        time.Time `json:"start_time"`
+	ClientRemoteAddr string    `json:"client_remote_address"`
+	ReqPath          string    `json:"request_path"`
+	SnapshotTime     time.Time `json:"snapshot_time"`
+}
+
+func (c *Core) StoreInFlightReqData(reqID string, data *InFlightReqData) {
+	c.inFlightReqData.InFlightReqMap.Store(reqID, data)
+	c.inFlightReqData.InFlightReqCount.Inc()
+
+}
+
+func (c *Core) DeleteInFlightReqData(reqID string) {
+	c.inFlightReqData.InFlightReqMap.Delete(reqID)
+	c.inFlightReqData.InFlightReqCount.Dec()
+}
+
+func (c *Core) GetInFlightReqData() (map[string]*InFlightReqData, error) {
+
+	currentInFlightReqMap := make(map[string]*InFlightReqData)
+	syncMapRangeResult := true
+	c.inFlightReqData.InFlightReqMap.Range(func(key, value interface{}) bool {
+		v, ok := value.(*InFlightReqData)
+		if !ok {
```

Comment:
Is this a realistic failure scenario?  I approve of defensive programming, but given that there's only one writer to this map maybe we don't need to be so cautious here?